### ISSUE TITLE
Add diffie-hellman-group-exchange-sha256 key exchange algorithm to conch

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7672]
|[<img alt="cjwatson's avatar" src="https://avatars.githubusercontent.com/u/2429141?s=50" width="50" height="50">](https://github.com/cjwatson)| @cjwatson reported|
|-|-|
|Trac ID|trac#7672|
|Type|enhancement|
|Created|2014-10-09 13:00:37Z|
|Branch|https://github.com/twisted/twisted/tree/conch-ssh-dhgex-sha256-7672|

OpenSSH has recently removed support for pre-SHA2 key exchange algorithms, including diffie-hellman-{group-exchange,group1}-sha1.  This patch adds support for diffie-hellman-group-exchange-sha256 per RFC4419, allowing interoperation with OpenSSH 6.7.  It's mostly a straight cargo-cult from the -sha1 variant.

Unfortunately I haven't been able to get OpenSSH's interop tests working with conch as yet.  Looking at the test server log confirms that it's completing the key exchange and initiating a session, but the transferred test file seems to be truncated to zero bytes.  It would be good if somebody could look into this.  However, things seem to work interactively and twisted.conch's own tests pass with my modifications, so I'm reasonably confident that I've got the basics right.

Attachments:

* [dhgex-sha256.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a8e/a8e8a57904b45622d5098a2c7a15ef8069b3d435/2b64edb225f8f14739f3884396db0e2b9d2a0fe9.patch) (22640 bytes) - added by cjwatson on 2014-10-09 13:01:36Z - Add diffie-hellman-group-exchange-sha256
* [dhgex-sha256-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a8e/a8e8a57904b45622d5098a2c7a15ef8069b3d435/3d0033f395e3802dbf6e26fc4f19aa9f530dcc2e.patch) (36037 bytes) - added by cjwatson on 2015-10-23 16:03:16Z - 
* [dhgex-sha256-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a8e/a8e8a57904b45622d5098a2c7a15ef8069b3d435/e4799b0cdeab2b84d8cd32dd0f1834daa54387fa.patch) (37906 bytes) - added by cjwatson on 2015-11-09 13:12:29Z - 
* [dhgex-sha256-2-to-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a8e/a8e8a57904b45622d5098a2c7a15ef8069b3d435/1b3c38a13a5f2c8344d0b07b2911aecbbd2415b9.patch) (11013 bytes) - added by cjwatson on 2015-11-09 13:12:40Z - 
* [dhgex-sha256-4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a8e/a8e8a57904b45622d5098a2c7a15ef8069b3d435/dca91a5273c29e3286f5b3c2f2192061984e65a4.patch) (38664 bytes) - added by cjwatson on 2015-11-09 13:26:19Z - 
* [dhgex-sha256-3-to-4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a8e/a8e8a57904b45622d5098a2c7a15ef8069b3d435/2ead35b56de989a666064f489913de10979ded5a.patch) (758 bytes) - added by cjwatson on 2015-11-09 13:26:27Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7672 7672
type__enhancement enhancement
reporter__cjwatson cjwatson
priority__normal normal
milestone__ 
branch__branches_conch_ssh_dhgex_sha256_7672 branches/conch-ssh-dhgex-sha256-7672
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__None None
time__1412859637369017 1412859637369017
changetime__1447098603239421 1447098603239421
version__None None
owner__cjwatson cjwatson
cc__amtota cc__mkuron
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban set owner to @cjwatson|
|-|-|

Many thanks for the patch!

Please note that to have this merged, it will need a news file. https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles

Also, to help apply the patch it would be nice to configure git, not to create diffs with prefix:

https://twistedmatrix.com/trac/wiki/GitMirror#CreatingaGitclone


When I applied the patch on trunk, one hunk was reject (don't know why)... but I can figure it out and do the review with this diff

```
$ patch -p1 < ~/Downloads/dhgex-sha256.patch 
patching file twisted/conch/ssh/factory.py
patching file twisted/conch/ssh/transport.py
patching file twisted/conch/test/test_ssh.py
patching file twisted/conch/test/test_transport.py
Hunk #6 FAILED at 1367.
1 out of 13 hunks FAILED -- saving rejects to file twisted/conch/test/test_transport.py.rej

```

---------


Code from factory.py and test_ssh.py looks good.

------

For transport.py

I see this code in more places

```
if self.kexAlg == 'diffie-hellman-group-exchange-sha256':
  h = sha256()
else:
  h = sha1()
```

I suggest that in ssh_KEXINIT to also set a self.keyAlgHashMethod once the keyAlg is set and then just use this reference.


--------

test_transport.py

1. Rename test_getKey(self) to test_getKeySHA1(self)

2. Rename test_getKey_sha256 to test_getKeySHA256 ... sorry for asking this, but it looks like this is the way Twisted developers think that they can improve test readability... maybe we can ask them to reconsider this.

3. In test test_KEX_DH_GEX_REQUEST_sha256 and test_KEX_DH_GEX_INIT_after_REQUEST_sha256 (and similar) I see a lot of duplicate code, please extract it into a helper method.

--------

Here are my steps:


1. Got and compiled latest OpenSSH release openssh-6.7p1.tar.gz

2. Start a modified version of sshsimpleserver.py in which it loads primes from '/etc/ssh/moduli' . Here is the code https://gist.github.com/adiroiban/c0e309b6d58494b8ebea

3. Tried to connect using SSH client and forcing sha256 ```./ssh -oKexAlgorithms=diffie-hellman-group-exchange-sha256 -p 5022 user@localhost```

It works... but with a huge delay after ``` debug1: expecting SSH2_MSG_KEX_DH_GEX_REPLY ``` ... delay in which no new connections were accepted... so this is an CPU lock :(

Note that the delay was also present for sha1, so is not something new.

Did you also get the slow key exchange process? I got a i7-4500U CPU @ 1.80GHz 

This is not a problem for this ticket, but maybe we should create a new ticket to use deferToThread for computing the key exchange values.

-----------

I have not tested a SFTP server / client as there is no simple example in Twisted source code... I will brink this over the email list and see if this deserve a ticket

Meanwhile I will try to get the code for a simple SFTP server so that we can share with other reviewers of this ticket.


Many thanks for this patch!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|

Please take a look at latest development from [#7717](https://github.com/twisted/twisted/issues/7717)

**Comment by twisted-trac** [comment]
|[<img alt="cjwatson's avatar" src="https://avatars.githubusercontent.com/u/2429141?s=50" width="50" height="50">](https://github.com/cjwatson)<a name="note_7"></a>|@cjwatson set owner to @adiroiban|
|-|-|

I've attached an updated version of this patch, based on the work in [#7717](https://github.com/twisted/twisted/issues/7717).  This includes a fair amount of test refactoring to reduce duplication, which I'm afraid involved moving some methods around a bit, so that part of the diff is a little noisy.

Regarding your performance issue, this mostly goes away if you install gmpy as recommended in the conch documentation.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban commented|
|-|-|

(In [45918]) Branching to conch-ssh-dhgex-sha256-7672.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban set owner to @cjwatson|
|-|-|

Many thanks for you work!

Maybe we can ramame getHashAlgorithm into getHashProcessor or something to hint that this is a callable and not the hash algo name.

In `twisted/conch/ssh/factory.py` in buildProtocol can we have a dynamic lookup which removes all non-fixed algorithm names?


In twisted/conch/test/test_transport.py in getPrimes, can we update the docstring to inform about generic non-fixed key exchanges, rather than listing here each new method that we support?


In BaseSSHTransportTests there is a double space... I can also fix it before merge.

In ServerSSHTransportTests maybe we need to rename test_KEXINIT to test_KEXINITAllAlgorithms and update the docstrign that this is a kexinit for the case when all algorithms are supported.

Same for ClientSSHTransportTests.

------

Other than that, it looks look, and the test pass.

Please review my comments and resubmit.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="cjwatson's avatar" src="https://avatars.githubusercontent.com/u/2429141?s=50" width="50" height="50">](https://github.com/cjwatson)<a name="note_10"></a>|@cjwatson set owner to @adiroiban|
|-|-|

I've addressed your review comments, thanks.  It would be nice if the style guide documented whether to use dot-space or dot-space-space, since both are in widespread use in Twisted.

**Comment by twisted-trac** [comment]
|[<img alt="cjwatson's avatar" src="https://avatars.githubusercontent.com/u/2429141?s=50" width="50" height="50">](https://github.com/cjwatson)<a name="note_11"></a>|@cjwatson commented|
|-|-|

I missed an OpenSSH interoperability test (requires OpenSSH >= 4.4).  Fortunately it's pretty trivial.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban set owner to @cjwatson|
|-|-|

Thanks for the latest patches.

I did a few small changes while reviewing the tests and trying to check that the test docstring matches the test execution.

Please check my changes.

If all is ok I will merge.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="cjwatson's avatar" src="https://avatars.githubusercontent.com/u/2429141?s=50" width="50" height="50">](https://github.com/cjwatson)<a name="note_13"></a>|@cjwatson set owner to @adiroiban|
|-|-|

Your changes look fine, thanks, except that in one place in a comment you say "diffie-hellman-group1-sha256" where you should say "diffie-hellman-group-exchange-sha256".

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_14"></a>|@adiroiban set owner to @cjwatson|
|-|-|

Thanks. I have fixed it. Tests pass.  Will merge.

I have also run a manual test using the docs/conch/examples/sshsimpleserver.py

```
ssh -p 5022 user@localhost -oKexAlgorithms=diffie-hellman-group-exchange-sha256
```

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_15"></a>|@adiroiban set status to `closed`|
|-|-|

(In [46130]) Merge conch-ssh-dhgex-sha256-7672: Add diffie-hellman-group-exchange-sha256 to twisted.conch.ssh

Author: cjwatson
Reviewer: adiroiban
Fixes: [#7672](https://github.com/twisted/twisted/issues/7672)
